[DRAFT] SK assistant for agentchat implementation

lspinheiro opened the pull request:
<!-- Thank you for your contribution! Please review https://microsoft.github.io/autogen/docs/Contribute before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

Still need to update message conversion to handle non-text types, add docstring, etc. But the basics are thre.

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

ekzhu commented on python/packages/autogen-core/docs/src/reference/index.md:
```diff
@@ -46,6 +46,7 @@ python/autogen_ext.agents.web_surfer
 python/autogen_ext.agents.file_surfer
 python/autogen_ext.agents.video_surfer
 python/autogen_ext.agents.video_surfer.tools
+python/autogen_ext.agents.semantic_kernel
```
let's move this up right after the openai assistant.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/agents/semantic_kernel/_sk_assistant_agent.py:
```diff
@@ -0,0 +1,327 @@
+                if sk_message.content:
+                    partial_text = sk_message.content
+                    accumulated_reply.append(partial_text)
+                    yield TextMessage(content=partial_text, source=self.name)
```
To keep the behavior consistent with AssistantAgent, let's accumulate and output only complete messages rather than partial messages.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/agents/semantic_kernel/_sk_assistant_agent.py:
```diff
@@ -0,0 +1,318 @@
+            kernel=self._kernel,
+        )
+        # Convert SK's list of responses into a single final text
+        assistant_reply = "\n".join(r.content for r in sk_responses if r.content)
```
For SK's list of responses, are they separate messages or a single message? Is it possible for this to contain messages that may contain tool calls?

lspinheiro replied:
It returns a list

```py
    async def get_chat_message_contents(
        self,
        chat_history: "ChatHistory",
        settings: "PromptExecutionSettings",
        **kwargs: Any,
    ) -> list["ChatMessageContent"]:
 ```

But it can return tool calls, it is a good catch. This depends on the configuration of the prompt settings. In the model adapter we are forcing semantic kernel to return tool calls to keep the same contract of the chat completion client. Here in the sample I'm using semantic kernel to automatically execute the function but this is not being forced in the configuration. Do you think we should force it or allow it to return tool calls? We may need to be more explicit about this behavior in the docs and maybe log some warnings

lspinheiro replied:
@ekzhu , do you have an idea of how you want to handle function calls? The output is similar to what we would get from the client. We have some options.

1. Override the prompt setting to execute it
2. Execute it manually by calling the kernel function (need to look into it, but looks possible)
3.  Return a tool call message / handoff? (unsure how handoffs work and what type of agent configuration would be needed).

Thoughts?

lspinheiro replied:
And there is also a validation that would throw a value error with a descriptive message in case function call message is returned by the model client, but this is not an expected behavior now.

ekzhu replied:
Regarding to the response message being a concatenation of multiple messages. Do you think it should be a sequence of event messages (e.g., ToolCallRequestedEvent) followed by a final response, as in the `AssistantAgent`.

lspinheiro replied:
I asked Evan form the SK team and he said it actually only returns a single message in the list, the list format was added just in case. We could add an assert/exception handler to check if it is a single element and then update the code to only process the expected message. That could make the processing more explicit while providing a clear indication in case this expectation is broken in the future.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/agents/semantic_kernel/_sk_assistant_agent.py:
```diff
@@ -0,0 +1,342 @@
+    def __init__(
+        self,
+        name: str,
+        description: str,
```
Let's make this field also optional with a default value similar to the "AssistantAgent". 

Use an `*` to separate the positional argument and the keyword arguments with default values.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/agents/semantic_kernel/_sk_assistant_agent.py:
```diff
@@ -0,0 +1,342 @@
+        execution_settings (PromptExecutionSettings, optional):
+            Optional prompt execution settings to override defaults.
+
+    Example usage:
```
Requiring a Bing search API key adds barrier to entry. I think we should have another example before this one to show the usage of basic function-based tools like calculator. 

Also we need to reference relevant docs in Semantic Kernel documentation whenever we mention a new concept from SK.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/agents/semantic_kernel/_sk_assistant_agent.py:
```diff
@@ -0,0 +1,342 @@
+
+    @property
+    def produced_message_types(self) -> Sequence[type[ChatMessage]]:
+        return [TextMessage]
```
Looks like the agent can't produce `HandoffMessage`, in that case we need to make a TODO and reference an new issue. Similar to #5496

lspinheiro replied:
Are there any guides on how to implement handoffs? I can add support but I'm not familiar with the conditions for when handoffs should occur.

ekzhu replied:
As a future PR.

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/agents/semantic_kernel/_sk_assistant_agent.py:
```diff
@@ -0,0 +1,342 @@
+        ):
+            for sk_message in sk_message_list:
+                # Check for function calls
+                if any(isinstance(item, FunctionCallContent) for item in sk_message.items):
```
I am not sure if we should raise here. Because even though the execution setting is auto, we can still get function call content as part of the message list.

lspinheiro replied:
I added the exception there because it shouldn't happen based on the constraints. If the execution setting is auto, semantic kernel handles function call content internally. It will cause an exception in either case because the code is not prepared to properly process the function call content, I just thought an explicit message would be easier to debug.

ekzhu replied:
I tried the sample in the API doc and actually got this error when I used OpenAI model.

jsburckhardt commented:
hi folks - @lspinheiro this is exactly the feature we are after - is there anyway I can help to bring this one forward? Cheers!

lspinheiro commented:
> hi folks - @lspinheiro this is exactly the feature we are after - is there anyway I can help to bring this one forward? Cheers!

Hi @jsburckhardt , I'm just tidying up the PR and hoping to get it ready for review again today. If you are familiar with semantic kernel, feedback on the implementation would be a great way to help.

lspinheiro commented:
@ekzhu I believe all comments have been addressed.

1. We are now emitting agent events for function calls in streaming
2. Added a new example using a custom plugin that doesnt depend on the bing api key. Also added some refs to SK docs.
3. I added a comment about the handoff to avoid it being a blocker, I will have a look into it and see if I can implement it both to the openai assistant agent and to sk assistant agent.

lspinheiro commented:
@ekzhu @jackgerrits , I believe this is ready for use. It would be nice to get this merged before it gets stale. Are you able to have a look or suggest someone to review?

ekzhu commented:
Will get to it. Though we seem to need to reconsider? It maybe that the wrapper on actual agent is what people want

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/agents/semantic_kernel/_sk_assistant_agent.py:
```diff
@@ -0,0 +1,528 @@
+        sk_response = sk_responses[0]
+
+        for item in sk_response.items:
+            if isinstance(item, TextContent):
```
What do we do with the other content types?

ekzhu commented on python/packages/autogen-ext/src/autogen_ext/agents/semantic_kernel/_sk_assistant_agent.py:
```diff
@@ -0,0 +1,528 @@
+                            tool_call_result = FunctionExecutionResult(
+                                content=exec_results,
+                                call_id=last_function_call_id,
+                                is_error=False,
```
How does SK handles exceptions raised from function calls? Is there a way we can determine if the result is an error?

lpinheiroms replied:
It seems the function error is propagated to the model. There isnt a specific return type for function execution error and the exception is not raised either. That is the feedback I got from Evan from the SK team.
